### PR TITLE
Add support for Zyxel Keenetic Viva router

### PR DESCRIPTION
**Pull request by 13hakta**
Implemented support for MDIO with Mediatek internal switch.
Added initial LED management support for RTL8367R/VB.
Added support for MIB in MT7620.

### CODE REVIEW

**Comment by miltador** [comment]
Signed-off-by is missing in commit messages.

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 1]
```diff
@@ -0,0 +1,280 @@
+Index: linux-4.4.27/drivers/net/ethernet/mediatek/gsw_mt7620.c
+===================================================================
+--- linux-4.4.27.orig/drivers/net/ethernet/mediatek/gsw_mt7620.c
++++ linux-4.4.27/drivers/net/ethernet/mediatek/gsw_mt7620.c
+@@ -67,6 +67,9 @@ static void mt7620_hw_init(struct mt7620
+ 	rt_sysc_w32(rt_sysc_r32(SYSC_REG_CFG1) | BIT(8), SYSC_REG_CFG1);
+ 	mtk_switch_w32(gsw, mtk_switch_r32(gsw, GSW_REG_CKGCR) & ~(0x3 << 4), GSW_REG_CKGCR);
+ 
++	// Enable MIB stats
++	mtk_switch_w32(gsw, mtk_switch_r32(gsw, GSW_REG_MIB_CNT_EN) | (1 << 1), GSW_REG_MIB_CNT_EN);
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: use c style comments

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+Index: linux-4.4.27/drivers/net/ethernet/mediatek/gsw_mt7620.h
+===================================================================
+--- linux-4.4.27.orig/drivers/net/ethernet/mediatek/gsw_mt7620.h
++++ linux-4.4.27/drivers/net/ethernet/mediatek/gsw_mt7620.h
+@@ -76,6 +76,8 @@
+ #define PHY_PRE_EN		BIT(30)
+ #define PMY_MDC_CONF(_x)	((_x & 0x3f) << 24)
+ 
++#define	GSW_REG_MIB_CNT_EN	0x4000
++
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: bad indenting, wrong position inside the file. add 0x4000 in numerical order with the other defines

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  * Copyright (C) 2013 John Crispin <blogic@openwrt.org>
++ * Copyright (C) 2016 Vitaly Chekryzhev <13hakta@gmail.com>
+  */
+ 
+ #include <linux/if.h>
+@@ -76,6 +77,95 @@ struct mt7xxx_mib_desc {
+ 	const char *name;
+ };
+ 
++#define MT7620_MIB_COUNTER_BASE	0x1010
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: defines go in the header file
dont use comments at the end of a line
dont use c++ comments

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+ {
+ 	unsigned int port_base;
+@@ -683,6 +761,111 @@ static int mt7621_sw_get_port_mib(struct
+ 	return 0;
+ }
+ 
++static u64 get_mib_counter_7620(struct mt7530_priv *priv, int i)
++{
++	u64 t;
++
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: the other is called hi so call this one lo

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
++	t = mt7530_r32(priv, MT7620_MIB_COUNTER_BASE + mt7620_mibs[i].offset);
++	if (mt7620_mibs[i].size == 2) {
++		u64 hi;
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: none of the defined MIBS use 2, why have the code for it here ?

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
++	return t;
++}
++
++static u64 get_mib_counter_port_7620(struct mt7530_priv *priv, int i, int port)
++{
++	unsigned int port_base;
++	u64 t;
++
++	port_base = MT7620_MIB_COUNTER_BASE_PORT +
++		    MT7620_MIB_COUNTER_PORT_OFFSET * port;
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: a;though not strictly required, please add () around the multiplication

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
++	}
++
++	return t;
++}
++
++static int mt7530_sw_get_mib(struct switch_dev *dev,
++				  const struct switch_attr *attr,
++				  struct switch_val *val)
++{
++	static char buf[4096];
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: no, no, no ... though shalt not put 4k on the stack

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
++	struct mt7530_priv *priv = container_of(dev, struct mt7530_priv, swdev);
++	int i, len = 0;
++
++	len += snprintf(buf + len, sizeof(buf) - len, "Switch MIB counters\n");
++
++	for (i = 0; i < sizeof(mt7620_mibs) / sizeof(*mt7620_mibs); ++i) {
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: the kernel has a macro called ARRAY_SIZE

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch [hunk 9]
```diff
@@ -0,0 +1,280 @@
...
++static const struct switch_attr mt7530_global[] = {
++	{
++		.type = SWITCH_TYPE_INT,
++		.name = "enable_vlan",
++		.description = "VLAN mode (1:enabled)",
++		.max = 1,
++		.id = MT7530_ATTR_ENABLE_VLAN,
++		.get = mt7530_get_vlan_enable,
++		.set = mt7530_set_vlan_enable,
++	}, {
```
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: the enable_vlan part is unrelated

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch (hunk 7 above)
Comment (13hakta) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: How could it be accepted to openwrt? That code was just adapted from neighbour method.

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch (hunk 5 above)
Comment (13hakta) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: inherited from neighbour method

Comment (13hakta) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: there are few of them for 7621 mibs.

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch (hunk 9 above)
Comment (13hakta) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: that was untouched part of existing code.

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch (hunk 5 above)
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: ok

File: target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch (hunk 9 above)
Comment (blogic) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: oh i see you are just moving the code to a different location int he file ?

Comment (13hakta) [inline target/linux/ramips/patches-4.4/0701-mt7620-add-mib-stats.patch]: some parts need refactoring, made relocation part of code for new methods to be visible for further usage.
